monix-tail/shared/src/main/scala/monix/tail/Iterant.scala

```diff
@@ -2125,6 +2132,16 @@ object Iterant extends IterantInstances {
   def range[F[_]](from: Int, until: Int, step: Int = 1)(implicit F: Applicative[F]): Iterant[F, Int] =
     NextBatch(Batch.range(from, until, step), F.pure(empty[F, Int]), F.unit)
 
+  /** Builds a stream that repeats the items provided in argument.
+    *
+    * It terminates either on error or if the source is empty.
+    */
+  def repeat[F[_], A](elems: A*)(implicit F: Sync[F]): Iterant[F, A] = elems match {
+    case Seq() => Iterant.empty
+    case Seq(elem) => Next[F, A](elem, F.delay(repeat(elem)), F.unit)
+    case _ => NextBatch[F, A](Batch(elems: _*), F.delay(repeat(elems: _*)), F.unit)
```

Comment:
It should be possible to optimize away the recursion / allocation on each step like so:
```scala
var result: Iterant[F, A] = null
result = NextBatch(Batch(elems: _*), F.delay(result), F.unit)
result
```
(the trick is not mine, [here's](https://japgolly.blogspot.com.by/2017/12/practical-awesome-recursion-ch-02.html) where I found it)

Reply:
It will be desugared to something along the lines of
```scala
val result: ObjectRef[Iterant[F, A]] = ObjectRef.create(null)
// delay accepts by-name parameter, which is Function0 at runtime
result.elem = Next[F, A](elem, F.delay(() => result.elem), F.unit)
result.elem
```

The null will be observable if, and only if by-name parameter passed to delay evaluates *before* the assignment happens on the same line. It will happen if you're using `pure`, which is strict in `cats`, but for `delay` this would be a violation of `unsequencedDelayIsNoop` law from `Sync`, so it cannot happen (anything goes if the instance is not lawful, no need to support this).

In the article, the assignment happens to function, for which body, too, will never be evaluated until the function called, and the call only happens after the variable was assigned.